### PR TITLE
Add FalconGenerator and all the needed machinery

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This change adds support for generation using the Falcon models. The generator wrapper is FalconGenerator. Falcon uses the Refined Web Model, which is slightly different from other models that we currently have:
    
- The model can use shared key and value heads. This is implemented using the new qkv_head_sharing argument of the self attention layer along with the QkvHeadSharing enum.
- The model can use all linear layers without bias. This is accommodated by adding a new use_bias argument to the self attention layer and the pointwise feed-forward layer.
- The activation is always GELU.
    
Unfortunately, we cannot test decoding with cache, position tensors, or the causal LM against the upstream implementation, since it is broken in two ways:
    
- The upstream implementation uses the PyTorch scaled dot product attention, which creates invalid causal attention masks. When the query and key shapes differ.
- The upstream implementation does not take into account that rotary embeddings should be offset for position when using a cache. As a result, the new query and key pieces will always be treated as starting at position 0.
    
I still need to investigate the differences between the NeoX tokenizer and the Refined Web Model tokenizer. I'll address this in a future PR.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/attention.py [hunk 1]
```diff
@@ -73,6 +73,16 @@ def apply_causal_mask(input: Tensor) -> Tensor:
     return torch.where(causal_mask, input, blocked_value)
 
 
+class QkvHeadSharing(IntEnum):
+    """Sharing of head parameters."""
+
+    #: No parameters are shared between heads.
+    NONE = 0
+
+    #: Key shares heads, value shares heads, query has separate heads.
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:82]:
```suggestion
    #: Key and value share heads, query has separate heads.
```

File: curated_transformers/models/attention.py [hunk 2]
```diff
@@ -224,18 +237,40 @@ def __init__(
...
+                self.model_dim, self.model_dim, bias=use_bias, device=device
+            )
+            self.key = Linear(self.model_dim, kv_dim, bias=use_bias, device=device)
+            self.value = Linear(self.model_dim, kv_dim, bias=use_bias, device=device)
         else:
             self.input = Linear(
                 self.model_dim,
-                self.model_dim * 3,
+                self.model_dim + 2 * kv_dim,
+                bias=use_bias,
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:267]: Was the absence of `bias` an oversight until now or is it required for Falcon?

File: curated_transformers/models/feedforward.py [hunk 3]
```diff
@@ -31,12 +31,17 @@ def __init__(
         hidden_act: str = "gelu",
         hidden_width: int = 768,
         intermediate_width: int = 3072,
+        use_bias: bool,
```
Comment (shadeMe) [inline curated_transformers/models/feedforward.py:34]: We don't want a default here?

File: curated_transformers/models/attention.py [hunk 4]
```diff
@@ -73,6 +73,16 @@ def apply_causal_mask(input: Tensor) -> Tensor:
...
 
 
+class QkvHeadSharing(IntEnum):
+    """Sharing of head parameters."""
+
+    #: No parameters are shared between heads.
+    NONE = 0
+
+    #: Key shares heads, value shares heads, query has separate heads.
+    KEY_VALUE = 1
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:83]: We could also mention [the paper](https://arxiv.org/abs/1911.02150) here.

File: curated_transformers/models/refined_web_model/causal_lm.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        :param config: Model configuration.
+        :param device: Device on which the module is to be initialized.
+        """
+        super().__init__()
+
+        self.decoder = RefinedWebModelDecoder(config, device=device)
+        self.output_embeddings = Linear(
+            in_features=config.layer.hidden_width,
+            out_features=config.embedding.vocab_size,
+            bias=False,
```
Comment (shadeMe) [inline curated_transformers/models/refined_web_model/causal_lm.py:35]: No `bias` term for the LM head (because of the layernorm that's performed on the last layer's output)?

File: curated_transformers/models/refined_web_model/config.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        use_bias: bool = False,
+    ):
+        """
+        :param dropout_prob: Dropout to apply after attention.
+        :param hidden_width: Hidden width of the transformer.
+        :param multi_query: Use multiple query heads and single key and
+            value heads.
+        :param num_attention_heads: Number of attention heads.
+        :param rotary_base: Base in signifying the rotary embedding period.
+        :param rotary_fraction: fraction of hidden width to apply rotary
```
Comment (shadeMe) [inline curated_transformers/models/refined_web_model/config.py:34]:
```suggestion
        :param rotary_fraction: Fraction of hidden width to apply rotary
```

File: curated_transformers/models/refined_web_model/config.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            embeddings to. Must be in [0,1].
+        :param use_bias: use bias in linear layers.
```
Comment (shadeMe) [inline curated_transformers/models/refined_web_model/config.py:36]:
```suggestion
        :param use_bias: Use bias in linear layers.
```

File: curated_transformers/models/refined_web_model/config.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        num_hidden_layers: int = 32,
+        layer_norm_eps: float = 1e-5,
+        use_bias: bool = False,
+    ) -> None:
+        """
+        :param dropout_prob: Dropout to apply after hidden layers.
+        :param hidden_width: Hidden width of the transformer.
+        :param layer_norm_eps: Epsilon for layer normalization.
+        :param num_hidden_layers: Number of hidden layers.
+        :param use_bias: use bias in linear layers.
```
Comment (shadeMe) [inline curated_transformers/models/refined_web_model/config.py:102]:
```suggestion
        :param use_bias: Use bias in linear layers.
```

File: curated_transformers/models/refined_web_model/config.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+            embedding layers.
+        :param hidden_width: Hidden width of the transformer.
+        :param layer_norm_eps: Epsilon for layer normalization.
+        :param multi_query: Use multiple query heads and single key and
+            value heads.
+        :param num_hidden_layers: Number of hidden layers.
+        :param rotary_embedding_base: Base in signifying the rotary embedding period.
+        :param rotary_embedding_fraction: Fraction of hidden width to apply rotary
+            embeddings to. Must be in [0,1].
+        :param use_bias: use bias in linear layers.
```
Comment (shadeMe) [inline curated_transformers/models/refined_web_model/config.py:146]:
```suggestion
        :param use_bias: Use bias in linear layers.
```

File: curated_transformers/models/refined_web_model/layer.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    def forward(
+        self,
+        x: Tensor,
+        attention_mask: Optional[AttentionMask],
+        cache: Optional[KeyValueCache] = None,
+        positions: Optional[Tensor] = None,
+        store_cache: bool = False,
+    ) -> Tuple[Tensor, Optional[KeyValueCache]]:
+        """
+        Apply the GPT-NeoX layer to the given piece hidden representations.
```
Comment (shadeMe) [inline curated_transformers/models/refined_web_model/layer.py:74]:
```suggestion
        Apply the Refined Web Model layer to the given piece hidden representations.
```

File: curated_transformers/models/refined_web_model/layer.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            look up rotary embeddings. Normally, these positions are calculated
+            automatically. But if the positions deviate for some reason, they
+            can be provided through this argument.
+        :param store_cache: Whether to cache the key/value representations for
+            future reuse.
+        :returns: Layer output.
+
+        Shapes:
+            x - (batch, seq_len, width)
+            attention_mask - (batch, seq_len)
```
Comment (shadeMe) [inline curated_transformers/models/refined_web_model/layer.py:92]: We should move the shape definitions to each parameter's docstring.

File: curated_transformers/tests/generation/test_falcon.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+
+@pytest.mark.veryslow
+@pytest.mark.skipif(not GPU_TESTS_ENABLED, reason="requires GPU")
+def test_generate_deterministic(falcon_generator):
+    prompts = [
+        "What is the Rust programming language?",
+        "What is spaCy?",
+    ]
+    assert falcon_generator(prompts, config=GreedyGeneratorConfig()) == [
```
Comment (shadeMe) [inline curated_transformers/tests/generation/test_falcon.py:27]: The test will look less noisy (and more debuggable) if we extracted the expected outputs to a variable and compared the outputs with it (`expected = expected[::-1]` for the reversed version).

File: curated_transformers/models/attention.py (hunk 2 above)
Comment (danieldk) [reply curated_transformers/models/attention.py:267]: `bias` is enabled by default. Falcon is the first model that we support that doesn't use bias for linear layers.

File: curated_transformers/models/feedforward.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/models/feedforward.py:34]: I think we should do the opposite here and remove all the defaults (like attention). I'll add a commit to do this.

File: curated_transformers/models/refined_web_model/causal_lm.py (hunk 5 above)
Comment (danieldk) [reply curated_transformers/models/refined_web_model/causal_lm.py:35]: It's common to not have bias on the output embeddings on the output. I don't really like the LM head terminology, it's really more like: the model produces an embedding and you then do a matrix multiplication between the predicted embedding and the output embedding matrix to find the token for which the output embedding is the closest to the predicted embedding.

File: curated_transformers/models/attention.py (hunk 1 above)
Comment (danieldk) [reply curated_transformers/models/attention.py:82]: I intentionally formulated it like this, otherwise it reads like the key and the value literally use the same head (v = k).

File: curated_transformers/models/feedforward.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/models/feedforward.py:27]: Removed all the defaults.

File: curated_transformers/models/attention.py (hunk 1 above)
Comment (shadeMe) [reply curated_transformers/models/attention.py:82]: Ah, I see.

File: curated_transformers/tests/generation/test_falcon.py (hunk 12 above)
Comment (danieldk) [reply curated_transformers/tests/generation/test_falcon.py]: That's a great idea! Changed the Dolly tests in the same way.

File: curated_transformers/models/refined_web_model/layer.py (hunk 11 above)
Comment (danieldk) [reply curated_transformers/models/refined_web_model/layer.py]: Done. Also in the decoder module.

File: curated_transformers/models/attention.py (hunk 4 above)
Comment (danieldk) [reply curated_transformers/models/attention.py:84]: Added.

File: curated_transformers/models/feedforward.py [hunk 13]
```diff
@@ -16,24 +16,24 @@ class PointwiseFeedForward(Module):
...
-        hidden_width: int = 768,
-        intermediate_width: int = 3072,
+        hidden_act: str,
+        hidden_width: int,
+        intermediate_width: int,
         use_bias: bool,
         device: Optional[torch.device] = None,
     ):
+        """
+        :param hidden_act: the activation function to apply, one of: "relu",
```
Comment (shadeMe) [inline curated_transformers/models/feedforward.py:31]:
```suggestion
        :param hidden_act: The activation function to apply, one of: "relu",
```

File: curated_transformers/models/feedforward.py [hunk 14]
```diff
@@ -16,24 +16,24 @@ class PointwiseFeedForward(Module):
...
+            "gelu" or "gelu_new".
+        :param hidden_width: the input and output width of the layer.
```
Comment (shadeMe) [inline curated_transformers/models/feedforward.py:33]:
```suggestion
        :param hidden_width: The input and output width of the layer.
```

File: curated_transformers/models/feedforward.py [hunk 15]
```diff
@@ -16,24 +16,24 @@ class PointwiseFeedForward(Module):
...
+        :param intermediate_width: the width of the projection to which the
```
Comment (shadeMe) [inline curated_transformers/models/feedforward.py:34]:
```suggestion
        :param intermediate_width: The width of the projection to which the
```

**Review by shadeMe — APPROVED** [review]
LGTM! Just some minor typos.
